### ISSUE TITLE
SwarmSpawner unable to contact notebook container when container name has upper case characters

### ISSUE THREAD

**Issue by hpbieker** [jupyterhub/dockerspawner#262]
I am using Keycloak for auth, and it gives user names that contains lower case, upper case, and numbers. By default, SwarmSpawner (and DockerSpawner) creates services (or containers) with a name in the format "<prefix>-<username>". 

The swarm load balancer is unable to do DNS lookup services that are upper case (https://github.com/moby/moby/issues/21169 / https://github.com/docker/libnetwork/issues/993) and will just hang.

It would be nice if:
* The username was automatically changed to lower case
* The container_name was automatically changed to lower case
* An error was reported if the container name ended up as an invalid name
* If docker supported DNS lookup of services in upper case...

**Comment by minrk** [comment]
Which Authenticator class & config are you using? Authenticators have a `.normalize_username` method which can do things like cast-lowercase. Indeed, it is the [default behavior](https://github.com/jupyterhub/jupyterhub/blob/0.9.4/jupyterhub/auth.py#L197) to do that in most authenticators, but you have to be sure that casting to lowercase cannot result in collisions (MediaWiki being one notable example where Min and min can be different users).

DockerSpawner should indeed be ensuring that containers get valid names. It can do this with escapes, but to be safe from potential collisions, it would have to use a safe escaping mechanism, which would result in `Abc` turning into something like `-41bc` rather than `abc`. The very least we could do is raise a more informative error.

**Comment by hpbieker** [comment]
I used the SwarmSpawner and OauthAuthenticator. We also use KeyCloak which generates these user names.

**Comment by cblomart** [comment]
I tested with Oauthenticator (azure) and SwarmSpawner. As we have long names in the directory, i moved to using the unique_name claim which will basicaly provide the mail address of the user.

The mail addresses are ``{firstname}.{lastname}@{domain}.{ext}``.

The service names provided by SwarmSpawner are in the form ``jupyter-{firstname}_2E{lastname}_40{domain}_2E{ext}``.

The jupyter hub registers to the proxy the name in the following form ``jupyter-{firstname}_2e{lastname}_40{domain}_2e{ext}``.

Thus the resulting service cannot be resolved: the names are not the same.

Obviously someone is lowering usernames somwhere... but SwarmSpwaner is not.

Looking at my jupyter logs:
```text
[I 2019-10-16 21:06:30.347 JupyterHub dockerspawner:1013] Starting service jupyter-{firstname}_2E{lastname}_40{domain}_2E{ext} (id: iywqxrs),
[I 2019-10-16 21:06:31.068 JupyterHub log:174] 302 GET /hub/spawn -> /hub/spawn-pending/jupyter-{firstname}.{lastname}@{domain}.{ext} ({firstname}.{lastname}@{domain}.{ext}@#.#.#.#) 1025.03ms,
[I 2019-10-16 21:06:31.174 JupyterHub pages:303] {firstname}.{lastname}@{domain}.{ext} is pending spawn,
[I 2019-10-16 21:06:31.210 JupyterHub log:174] 200 GET /hub/spawn-pending/{firstname}.{lastname}@{domain}.{ext} ({firstname}.{lastname}@{domain}.{ext}@#.#.#.#) 91.02ms,
[W 2019-10-16 21:06:35.975 JupyterHub _version:34] Single-user server has no version header, which means it is likely < 0.8. Expected 1.0.0,
[I 2019-10-16 21:06:35.976 JupyterHub base:810] User {firstname}.{lastname}@{domain}.{ext} took 5.916 seconds to start,
[I 2019-10-16 21:06:35.977 JupyterHub proxy:261] Adding user {firstname}.{lastname}@{domain}.{ext} to proxy /user/{firstname}.{lastname}@{domain}.{ext}/ => http://{firstname}_2e{lastname}_40{domain}_2e{ext}:8888
```
From jupyterhub [proxy.py:261](https://github.com/jupyterhub/jupyterhub/blob/5cbf9399b2dea3f65a1cb4faec648d5ed931e2f3/jupyterhub/proxy.py#L261): 
the hostname provided to proxy comes from ``spawner.server.host``

Looking in the sqlite database confirms it: the name is lowercase in there ```http://{firstname}_2e{lastname}_40{domain}_2e{ext}:8888``.

Would it be an possibility to let the jupyter admin decide to lowercase the names or not ?
(here i would gladly do this)

**Comment by cblomart** [comment]
I added an option to lowercase the object_name and proposed a pull request #342

**Comment by minrk** [comment]
This will be fixed by #262
